xds/internal/clients/grpctransport/grpc_transport_test.go

```diff
@@ -0,0 +1,301 @@
+// new client stream for the server.
+func (s) TestNewStream(t *testing.T) {
+	ts := setupTestServer(t)
+
+	tests := []struct {
+		name      string
+		serverURI string
+		wantErr   bool
+	}{
+		{
+			name:      "success",
+			serverURI: ts.lis.Addr().String(),
+			wantErr:   false,
+		},
+		{
+			name:      "error",
+			serverURI: "invalid-server-uri",
+			wantErr:   true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			serverCfg := clients.ServerConfig{
+				ServerURI:  test.serverURI,
+				Extensions: ServerConfigExtension{Credentials: insecure.NewBundle()},
+			}
+			builder := Builder{}
+			transport, err := builder.Build(serverCfg)
+			if err != nil {
+				t.Fatalf("failed to build transport: %v", err)
+			}
+			defer transport.Close()
+
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+			defer cancel()
+			_, err = transport.NewStream(ctx, "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources")
+			if (err != nil) != test.wantErr {
+				t.Fatalf("transport.NewStream() error = %v, wantErr %v", err, test.wantErr)
+			}
+		})
+	}
+}
+
+// TestStream_SendAndRecv verifies that grpcTransport.Stream.Send()
+// and grpcTransport.Stream.Recv() successfully send and receive messages
+// on the stream to and from the gRPC server.
+//
+// It starts a gRPC test server using setupTestServer(). The test then sends a
+// testDiscoverRequest on the stream and verifies that the received discovery
+// on the server is same as sent. It then wait to receive a
+// testDiscoverResponse from the server and verifies that the received
+// discovery response is same as sent from the server.
+func (s) TestStream_SendAndRecv(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout*2000)
+	defer cancel()
+
+	ts := setupTestServer(t)
+
+	// Build a grpc-based transport to the above server.
+	serverCfg := clients.ServerConfig{
+		ServerURI:  ts.lis.Addr().String(),
+		Extensions: ServerConfigExtension{Credentials: insecure.NewBundle()},
+	}
+	builder := Builder{}
+	transport, err := builder.Build(serverCfg)
+	if err != nil {
+		t.Fatalf("failed to build transport: %v", err)
+	}
+	defer transport.Close()
+
+	// Create a new stream to the server.
+	stream, err := transport.NewStream(ctx, "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources")
+	if err != nil {
+		t.Fatalf("failed to create stream: %v", err)
+	}
+
+	// Send a discovery request message on the stream.
+	msg, err := proto.Marshal(testDiscoverRequest)
+	if err != nil {
+		t.Fatalf("failed to marshal DiscoveryRequest: %v", err)
+	}
+	if err := stream.Send(msg); err != nil {
+		t.Fatalf("failed to send message: %v", err)
+	}
+
+	// Verify that the DiscoveryRequest received on the server was same as
+	// sent.
+	select {
+	case gotReq, ok := <-ts.requestChan:
+		if !ok {
+			t.Fatalf("ts.requestChan is closed")
+		}
```

Comment:
`requestChan` is never closed. I would get rid of this to simplify the test logic.